apps/omg/lib/omg/state/transaction/validator.ex

```diff
@@ -36,27 +37,71 @@ defmodule OMG.State.Transaction.Validator do
           | :unauthorized_spend
           | :utxo_not_found
 
-  @spec can_apply_spend(state :: Core.t(), tx :: Transaction.Recovered.t(), fees :: Fees.optional_fee_t()) ::
-          true | {{:error, exec_error()}, Core.t()}
-  def can_apply_spend(
-        %Core{utxos: utxos} = state,
-        %Transaction.Recovered{signed_tx: %{raw_tx: raw_tx}, witnesses: witnesses} = tx,
+  @type fee_claim_error :: :claiming_unsupported_token | :claiming_more_than_collected
+
+  @type process_error :: exec_error | fee_claim_error
+
+  @doc """
+  can process transaction?
+  """
+  @spec can_process(state :: Core.t(), tx :: Transaction.Recovered.t(), fees :: Fees.optional_fee_t()) ::
+          {:ok, :apply_spend, map()} | {:ok, :claim_fees, map()} | {{:error, process_error()}, Core.t()}
+  def can_process(
+        %Core{} = state,
+        %Transaction.Recovered{signed_tx: %{raw_tx: raw_tx}} = tx,
         fees
       ) do
+    case raw_tx do
+      %Transaction.Payment{} ->
+        can_apply_spend(state, tx, fees)
+
+      %Transaction.FeeTokenClaim{} ->
+        can_claim_fees(state, tx)
+    end
+  end
+
+  @spec can_apply_spend(state :: Core.t(), tx :: Transaction.Recovered.t(), fees :: Fees.optional_fee_t()) ::
+          {:ok, :apply_spend, map()} | {{:error, exec_error()}, Core.t()}
+  defp can_apply_spend(
+         %Core{utxos: utxos} = state,
+         %Transaction.Recovered{signed_tx: %{raw_tx: raw_tx}, witnesses: witnesses} = tx,
+         fees
+       ) do
     inputs = Transaction.get_inputs(tx)
 
-    with :ok <- validate_block_size(state),
+    with true <- not state.fee_claiming_started || {:error, :payments_rejected_during_fee_claiming},
+         :ok <- validate_block_size(state),
          :ok <- inputs_not_from_future_block?(state, inputs),
          {:ok, outputs_spent} <- UtxoSet.get_by_inputs(utxos, inputs),
          :ok <- authorized?(outputs_spent, witnesses),
          {:ok, implicit_paid_fee_by_currency} <- Transaction.Protocol.can_apply?(raw_tx, outputs_spent),
          true <- Fees.covered?(implicit_paid_fee_by_currency, fees) || {:error, :fees_not_covered} do
-      true
+      {:ok, :apply_spend, implicit_paid_fee_by_currency}
+    else
+      {:error, _reason} = error -> {error, state}
+    end
+  end
+
+  @spec can_claim_fees(Core.t(), Transaction.Recovered.t()) ::
+          {:ok, :claim_fees} | {{:error, fee_claim_error()}, Core.t()}
+  defp can_claim_fees(
+         %Core{fee_claimer_address: owner, fees_paid: fees_paid} = state,
+         %Transaction.Recovered{signed_tx: %{raw_tx: fee_tx}}
+       ) do
+    with outputs = make_outputs(owner, fees_paid),
+         {:ok, claimed_fee_token} <- Transaction.Protocol.can_apply?(fee_tx, outputs) do
```

Comment:
oh, this looks like overloading of the return of `Transaction.Protocol.can_apply?` too much.